### ISSUE TITLE
Arbitrary Same-Origin Stored XSS on React + HTML Outputs

### ISSUE THREAD

**Issue by bliutech** [uclaacm/TeachLAFrontend#1082]
Initially disclosed to @mizlan privately. Opening this issue for easier tracking and documentation.

## Vulnerability
On https://editor.uclaacm.com, there is an arbitrary [Same-Origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) [Stored XSS](https://owasp.org/www-community/attacks/xss/#stored-xss-attacks) for outputs of the editor specifically for both React & HTML. The issue arrises from lack of sandboxing of the output while using [`srcdoc`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement/srcdoc) causing the embedded content to be on the same origin as the editor ([source code](https://github.com/uclaacm/TeachLAFrontend/blob/ace76273fd05b2c22985d822accc14a1769fc92c/src/components/Output/Output.js#L120)). The impact could be that an adversary is able to share a page with a payload which could exfiltrate private data about a user or lead to arbitrary writes or downloads.

## POC
**React:**
![image](https://github.com/uclaacm/TeachLAFrontend/assets/67720812/efff2bd5-8dea-46f2-93f2-d10163bff475)

**HTML:**
![image](https://github.com/uclaacm/TeachLAFrontend/assets/67720812/59e2e35e-a59c-471a-adc8-38353deb5ded)

## Fix
The solution to this is to introduce the [`sandbox` ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#sandbox) property to the output iframe. This can be done in the following [source code](https://github.com/uclaacm/TeachLAFrontend/blob/ace76273fd05b2c22985d822accc14a1769fc92c/src/components/Output/Output.js#L120) with a behavior of `sandbox="allow-scripts allow-modals"`. This would allow many of the current features to be maintained while allowing resources to be isolated by the [Same-Origin Policy](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy).

<img width="1502" alt="image" src="https://github.com/uclaacm/TeachLAFrontend/assets/67720812/c912f874-d285-41c0-92f8-651f40e3f6b5">

**Comment by mizlan** [comment]
What are some tests I can do after a fix to ensure I've resolved the problem? @bliutech 

Is it enough to check the `document.domain` property?

**Comment by bliutech** [comment]
Yes. If the origins are not the same, then the Same-Origin Policy applies which should address this issue. 👍

**Comment by bliutech** [comment]
Update. [CVE-2023-42267](https://cve.report/CVE-2023-42267) is reserved for this.
